src/zope/app/tree/adapters.py

```diff
@@ -75,54 +76,44 @@ def getId(self):
                     if parent.__name__]
         return '\\'.join(parents)
 
+@implementer(IChildObjects)
+@adapter(IReadContainer)
 class ContainerChildObjects(object):
-    implements(IChildObjects)
-    adapts(IReadContainer)
 
     def __init__(self, context):
         self.context = context
 
     def hasChildren(self):
         # make sure we check for access
         try:
-            lenght = bool(len(self.context))
-            if lenght > 0:
-                return True
-            else:
-                return False
-        except Unauthorized:
+            return bool(len(self.context))
+        except Unauthorized: # pragma: no cover
             return False
 
     def getChildObjects(self):
-        if self.hasChildren():
-            return self.context.values()
-        else:
-            return []
+        return list(self.context.values()) if self.hasChildren() else []
```

Comment:
~~I don't quite get why `list(self.context.values())` shouldn't work for both cases.~~

Oh, it's because `hasChildren()` handles permission errors.